pkg/lib/database/order.go

```diff
@@ -17,17 +18,116 @@ import (
 type Order interface {
 	Compare(a, b *domain.Details) int
 	AnystoreSort() query.Sort
+	UpdateOrderMap(depDetails []*domain.Details) (updated bool)
 }
 
 // ObjectStore interface is used to enrich filters
 type ObjectStore interface {
 	SpaceId() string
 	Query(q Query) (records []Record, err error)
 	QueryRaw(filters *Filters, limit int, offset int) ([]Record, error)
+	QueryIterate(q Query, proc func(details *domain.Details)) (err error)
 	GetRelationFormatByKey(key domain.RelationKey) (model.RelationFormat, error)
 	ListRelationOptions(relationKey domain.RelationKey) (options []*model.RelationOption, err error)
 }
 
+type OrderMap struct {
+	data map[string]*domain.Details // objectId -> orderId
+}
+
+func NewOrderMap(data map[string]*domain.Details) *OrderMap {
+	return &OrderMap{data: data}
+}
+
+func (m *OrderMap) BuildOrderByKey(key domain.RelationKey, ids ...string) string {
+	if m == nil || len(m.data) == 0 {
+		return ""
+	}
+
+	var result string
+	for _, id := range ids {
+		if details, ok := m.data[id]; ok {
+			result += details.GetString(key)
+		}
+	}
+
+	return result
+}
+
+// Update updates orders only for objects that exist in OrderMap
+func (m *OrderMap) Update(details []*domain.Details) (anyUpdated bool) {
+	if m == nil || len(m.data) == 0 {
+		return false
+	}
+	for _, det := range details {
+		id := det.GetString(bundle.RelationKeyId)
+		updated := false
+		existingDetails, found := m.data[id]
+		if !found {
+			continue
+		}
+
+		orderId := det.GetString(bundle.RelationKeyOrderId)
+		if existingDetails.GetString(bundle.RelationKeyOrderId) != orderId {
+			updated = true
+			existingDetails.SetString(bundle.RelationKeyOrderId, orderId)
+		}
+
+		name := det.GetString(bundle.RelationKeyName)
+		if existingDetails.GetString(bundle.RelationKeyName) != name {
+			updated = true
+			existingDetails.SetString(bundle.RelationKeyName, name)
+		}
+
+		if updated {
+			anyUpdated = true
+		}
+	}
+	return anyUpdated
+}
+
+// SetOrders sets order for ids that do not present in OrderMap
+func (m *OrderMap) SetOrders(store ObjectStore, ids ...string) {
+	if store == nil {
+		return
+	}
+
+	if m.data == nil {
+		m.data = make(map[string]*domain.Details, len(ids))
+	}
+
+	idsToSet := make([]string, 0, len(ids))
+	for _, id := range ids {
+		if _, found := m.data[id]; !found {
+			idsToSet = append(idsToSet, id)
+		}
+	}
+
+	if len(idsToSet) == 0 {
+		return
+	}
+
+	records, err := store.Query(Query{Filters: []FilterRequest{{
+		RelationKey: bundle.RelationKeyId,
+		Condition:   model.BlockContentDataviewFilter_In,
+		Value:       domain.StringList(idsToSet),
+	}}})
+	if err != nil {
+		log.Errorf("failed to query records to update orders: %v", err)
```

Comment:
Consider returning the error and log it on a some upper level. Function will become slightly cleaner